productcatalog/utils.bal

```diff
@@ -14,19 +14,34 @@
 // specific language governing permissions and limitations
 // under the License.
 
+type JsonProduct record {|
+    string id;
+    string name;
+    string description;
+    string picture;
+    record {|
+        string currencyCode;
+        int units;
+        int nanos;
+    |} priceUsd;
+    string[] categories;
+|};
+
 isolated function parseProductJson(json jsonContents) returns Product[]|error {
     json productsJson = check jsonContents.products;
     if productsJson !is json[] {
         return error("product array is not found");
     }
-    return from json productJson in productsJson
+
+    JsonProduct[] jsonProducts = check productsJson.fromJsonWithType();
+    return from JsonProduct jsonProduct in jsonProducts
```

Comment:
Use record de-structuring here. Then you don't need to refer to `JsonProduct `record and have field access. You can directly refer to record fields as variables in lines 39-44

Reply:
wanted to ask to use binding patterns here

Reply:
Updated in this
https://github.com/ballerina-guides/gcp-microservices-demo/pull/47/commits/4163ebf49228eb882b5b3f3a1b4edec8e1806151